_posts/2017-09-21-scalamacros.md

```diff
@@ -0,0 +1,438 @@
+analysis on "Loosing whiteboxity"
+
+> The main motivation for getting rid of whitebox expansion is simplification -
+> both of the macro expansion pipeline and the typechecker.  Currently, they
+> are inseparably intertwined, complicating both compiler evolution and tool
+> support.
+
+Note, however, that the portable design of macros v3 makes it possible to infer
+the correct result types for whitebox macros in IDEs such as IntelliJ.
+
+Quoting the [minutes from the Scala Center Advisory Board][SCP-014]:
+
+> Dotty, he [Martin Odersky] says, wants to be a “capable language” rather than
+> a “language toolbox”. So it matters whether whitebox macros are being used to do
+> “Scala-like” things, or to turn Scala into something else. So “we will have
+> to look at each one” of the ways whitebox macros are being used. 
+
+Adriaan Moors, the Scala compiler team lead at Lightbend agreed with Martin,
+and mentioned a current collaboration with Miles Sabin to improve scalac so
+that Shapeless and other libraries can rely less on macros and other
+nonstandard techniques
+
+If you want to see whitebox macros approved for inclusion in macros v3,
+we invite you to share your thoughts in [Scala Contributors].
+
+#### Macro annotations
+
+Macro annotations have neither been approved nor rejected for inclusion into
+macros v3.
+Macro annotations have the ability to synthesize publicly available definitions,
+accommodating the [public type provider] pattern.
+Popular macro annotation libraries include
+- [simulacrum](https://github.com/mpilquist/simulacrum) first-class syntax
+  support for type classes in Scala
+- [Freestyle](http://frees.io/): cohesive & pragmatic framework of FP centric
+  Scala libraries
+
+To show an example macro annotation, consider the following `@deriving` macro
+annotation from the library [Stalactite]
+
+```scala
+@deriving(Encoder, Decoder)
+case class User(name: String, age: Int)
+
+// expands into
+case class User(name: String, age: Int)
+object User {
+  implicit val encoder: Encoder[User] = DerivedEncoder.gen
+  implicit val decoder: Decoder[User] = DerivedDecoder.gen
+}
+```
+The unique feature of macro annotations is that they can synthesize publicly
+available definitions such as `User.encoder/decoder`.
+Neither whitebox or blackbox def macros have this capability.
+It is generally considered best practice to place implicit
+typeclass instances in the companion object.
+This pattern significantly improves on compile times and prevents code bloat
+compared to full-derivation at each call-site.
+
+Surely, it's possible to manually write out the
+`implicit val decoderBar: Decoder[Bar] = DerivedDecoder.gen`
+parts.
+However, for large business applications with many domain-specific data types
+and typeclasses, such boilerplate hurts readability and often falls prey to
+typos, leading to bugs.
+
+Old-fashioned code generation via scripting can used as an
+alternative to macro annotations.
```

Comment:
It may be interesting to discuss the trade-offs. For one, scripting doesn't provide access to ASTs.